File: desktopApp/src/jvmMain/kotlin/ui/composables/leftPanel/Cup.kt
package ui.composables.leftPanel

import androidx.compose.desktop.ui.tooling.preview.Preview
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.size
import androidx.compose.material.MaterialTheme
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.blur
import androidx.compose.ui.geometry.CornerRadius
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import themes.CupTheme

@Composable
fun Cup(height: Dp) {
    CupTheme {
        val primary = MaterialTheme.colors.primary
        val secondary = MaterialTheme.colors.secondary
        val onPrimary = MaterialTheme.colors.onPrimary

        Canvas(Modifier.size(height)) {
            val w = size.width
            val h = size.height
            val borderHeight = h / 7
            val bodyOffset = borderHeight / 3
            val bezierOffset = h / 80
            val bezierOffset2 = h / 200

            val topLeft = Offset(bodyOffset, borderHeight)
            val topRight = Offset(w - bodyOffset, borderHeight)
            val bottomRight = Offset(w - borderHeight * 3 / 2, h)
            val bottomLeft = Offset(borderHeight * 3 / 2, h)

            drawRoundRect(color = primary, size = Size(size.width, borderHeight), cornerRadius = CornerRadius(20f, 20f))
            drawLine(color = secondary, alpha = 0.5f, start = topLeft, end = topRight)
            drawPath(
                color = primary,
                path = Path().apply {
                    moveTo(topLeft.x, topLeft.y)
                    lineTo(topRight.x, topRight.y)

                    lineTo(bottomRight.x + bezierOffset2, bottomRight.y - bezierOffset)
                    cubicTo(
                        x1 = bottomRight.x + bezierOffset2, y1 = bottomRight.y - bezierOffset,
                        x2 = bottomRight.x, y2 = bottomRight.y,
                        x3 = bottomRight.x - bezierOffset, y3 = bottomRight.y
                    )
                    lineTo(bottomRight.x - bezierOffset, bottomRight.y)

                    lineTo(bottomLeft.x + bezierOffset, bottomLeft.y)
                    cubicTo(
                        x1 = bottomLeft.x + bezierOffset, y1 = bottomLeft.y,
                        x2 = bottomLeft.x, y2 = bottomLeft.y,
                        x3 = bottomLeft.x - bezierOffset2, y3 = bottomRight.y - bezierOffset
                    )
                    lineTo(bottomLeft.x - bezierOffset2, bottomRight.y - bezierOffset)
                }
            )
            drawPath(
                color = onPrimary,
                path = Path().apply {
                    moveTo((bottomLeft.x + topLeft.x) / 2, (bottomLeft.y + topLeft.y) / 2)
                    lineTo(bottomRight.x - bezierOffset, bottomRight.y)
                    lineTo(bottomLeft.x + bezierOffset, bottomLeft.y)
                    cubicTo(
                        x1 = bottomLeft.x + bezierOffset, y1 = bottomLeft.y,
                        x2 = bottomLeft.x, y2 = bottomLeft.y,
                        x3 = bottomLeft.x - bezierOffset2, y3 = bottomRight.y - bezierOffset
                    )
                    lineTo(bottomLeft.x - bezierOffset2, bottomRight.y - bezierOffset)
                }
            )
        }
    }
}

@Composable
@Preview
private fun CupPreview() {
    Box(
        Modifier.fillMaxSize().background(Color.LightGray),
        contentAlignment = Alignment.Center
    ) {
        Cup(50.dp)
    }
}

File: iosApp/iosApp/Views/Cup.swift
import SwiftUI

struct Cup: View {
    
    private var height: CGFloat
    
    init(_ size: CGFloat) {
        self.height = size
    }
    
    var body: some View {
        ThemeScope(theme: Themes.Cup()) { theme in
            ZStack(alignment: .top) {
                let w = height
                let h = height
                
                let borderHeight = h / 7
                let bodyOffset = borderHeight / 3
                let bezierOffset = h / 80
                let bezierOffset2 = h / 200
                
                let topLeft = CGPoint(x: bodyOffset, y: borderHeight)
                let topRight = CGPoint(x: w - bodyOffset, y: borderHeight)
                let bottomRight = CGPoint(x: w - borderHeight * 3 / 2, y: h)
                let bottomLeft = CGPoint(x: borderHeight * 3 / 2, y: h)
                
                RoundedRectangle(cornerRadius: 10)
                    .fill(theme.primary)
                    .frame(width: w, height: borderHeight)
                
                CupBodyShape(
                    topLeft: topLeft,
                    topRight: topRight,
                    bottomRight: bottomRight,
                    bottomLeft: bottomLeft,
                    bezierOffset: bezierOffset,
                    bezierOffset2: bezierOffset2
                ).fill(theme.primary)
                
                CupStripeShape(
                    topLeft: topLeft,
                    bottomRight: bottomRight,
                    bottomLeft: bottomLeft,
                    bezierOffset: bezierOffset,
                    bezierOffset2: bezierOffset2
                ).fill(theme.onPrimary)
            }
            .frame(width: height, height: height)
        }
    }
}

private struct CupBodyShape : Shape {
    
    private var topLeft: CGPoint
    private var topRight: CGPoint
    private var bottomRight: CGPoint
    private var bottomLeft: CGPoint
    private var bezierOffset: CGFloat
    private var bezierOffset2: CGFloat
    
    init(
        topLeft: CGPoint,
        topRight: CGPoint,
        bottomRight: CGPoint,
        bottomLeft: CGPoint,
        bezierOffset: CGFloat,
        bezierOffset2: CGFloat
    ) {
        self.topLeft = topLeft
        self.topRight = topRight
        self.bottomRight = bottomRight
        self.bottomLeft = bottomLeft
        self.bezierOffset = bezierOffset
        self.bezierOffset2 = bezierOffset2
    }
    
    func path(in rect: CGRect) -> Path {
        var path = Path()
        path.move(to: topLeft)
        path.addLine(to: topRight)
        path.addLine(to: CGPoint(x: bottomRight.x + bezierOffset2, y: bottomRight.y - bezierOffset))
        path.addQuadCurve(
            to: CGPoint(x: bottomRight.x - bezierOffset, y: bottomRight.y),
            control: bottomRight
        )
        path.addLine(to: CGPoint(x: bottomRight.x - bezierOffset, y: bottomRight.y))
        path.addLine(to: CGPoint(x: bottomLeft.x + bezierOffset, y: bottomLeft.y))
        path.addQuadCurve(
            to: CGPoint(x: bottomLeft.x - bezierOffset2, y: bottomLeft.y - bezierOffset),
            control: bottomLeft
        )
        path.addLine(to: CGPoint(x: bottomLeft.x - bezierOffset2, y: bottomRight.y - bezierOffset))
        return path
    }
}

private struct CupStripeShape : Shape {
    
    private var topLeft: CGPoint
    private var bottomRight: CGPoint
    private var bottomLeft: CGPoint
    private var bezierOffset: CGFloat
    private var bezierOffset2: CGFloat
    
    init(
        topLeft: CGPoint,
        bottomRight: CGPoint,
        bottomLeft: CGPoint,
        bezierOffset: CGFloat,
        bezierOffset2: CGFloat
    ) {
        self.topLeft = topLeft
        self.bottomRight = bottomRight
        self.bottomLeft = bottomLeft
        self.bezierOffset = bezierOffset
        self.bezierOffset2 = bezierOffset2
    }
    
    func path(in rect: CGRect) -> Path {
        var path = Path()
        path.move(to: CGPoint(x: (bottomLeft.x + topLeft.x) / 2, y: (bottomLeft.y + topLeft.y) / 2))
        path.addLine(to: CGPoint(x: bottomRight.x - bezierOffset, y: bottomRight.y))
        path.addLine(to: CGPoint(x: bottomLeft.x + bezierOffset, y: bottomLeft.y))
        path.addQuadCurve(
            to: CGPoint(x: bottomLeft.x - bezierOffset2, y: bottomRight.y - bezierOffset),
            control: bottomLeft
        )
        path.addLine(to: CGPoint(x: bottomLeft.x - bezierOffset2, y: bottomRight.y - bezierOffset))
        return path
    }
}

struct Cup_Previews: PreviewProvider {
    static var previews: some View {
        ZStack {
            Color.gray
            Cup(CGFloat(150))
        }
    }
}
